### PR TITLE
fast decimal scalars

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 1]
```diff
@@ -129,7 +132,14 @@ public static long absShort(@SqlType("decimal(p, s)") long arg)
         @SqlType("decimal(p, s)")
         public static Slice absLong(@SqlType("decimal(p, s)") Slice arg)
         {
-            return encodeUnscaledValue(decodeUnscaledValue(arg).abs());
+            if(isNegative(arg)) {
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: fix formatting

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 2]
```diff
@@ -698,82 +701,44 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(B
...
     {
-        long scale = context.getLiteral("num_scale");
-        long rescaleFactor = longTenToNth((int) scale);
-        return ImmutableList.of(rescaleFactor, scale);
-    }
+        @LiteralParameters({"p", "s", "rp", "rs"})
+        @SqlType("decimal(rp, rs)")
+        @Constraint(variable = "rp", expression = "min(38, p - s + min(1, s))")
+        @Constraint(variable = "rs", expression = "0")
+        public static long roundShort(@LiteralParameter("s") Long numScale, @SqlType("decimal(p, s)") long num)
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:703]: `Long` ?

Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:703]: rename `num`  abbreviation

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 3]
```diff
@@ -698,82 +701,44 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(B
...
-            return 0;
-        }
-        if (inputScale == 0) {
-            return num;
-        }
-        if (num < 0) {
-            return -roundShortDecimal(-num, rescaleFactor, inputScale);
+            long rescaleFactor = Decimals.longTenToNth(numScale.intValue());
+            long remainder = num % rescaleFactor;
+            long remainderBoundary = rescaleFactor >> 1;
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: `/ 2`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 4]
```diff
@@ -741,78 +738,69 @@ public static Slice roundLong(@LiteralParameter("s") Long numScale, @SqlType("de
...
-    }
+    @ScalarFunction("round")
+    @Description("round to nearest integer")
+    public static final class RoundN
+    {
+        @LiteralParameters({"p", "s", "rp"})
+        @SqlType("decimal(rp, s)")
+        @Constraint(variable = "rp", expression = "min(38, p + 1)")
+        public static long roundNShort(
+                @LiteralParameter("p") Long numPrecision,
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:743]: %s/num/value/g

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 5]
```diff
@@ -741,78 +738,69 @@ public static Slice roundLong(@LiteralParameter("s") Long numScale, @SqlType("de
...
+                @LiteralParameter("s") Long numScale,
+                @SqlType("decimal(p, s)") long num,
+                @SqlType(BIGINT) long decimals)
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:746]: `decimals`? Please rename to something meaningful like `decimalPlaces`. `decimals` is very missleading. Or maybe `returnScale`? Because I don't know from this code whether it is `decimalPlaces` or `returnScale`.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 6]
```diff
@@ -741,78 +738,69 @@ public static Slice roundLong(@LiteralParameter("s") Long numScale, @SqlType("de
...
-
-    public static Slice roundNShortLong(long num, long roundScale, long inputPrecision, long inputScale)
-    {
-        return roundNLongLong(encodeUnscaledValue(num), roundScale, inputPrecision, inputScale);
-    }
+        @LiteralParameters({"p", "s", "rp"})
+        @SqlType("decimal(rp, s)")
+        @Constraint(variable = "rp", expression = "min(38, p + 1)")
+        public static Slice roundNLong(
+                @LiteralParameter("s") Long numScale,
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:768]: `Long`?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 7]
```diff
@@ -686,156 +698,109 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(B
...
-    {
-        if (num == 0) {
-            return 0;
-        }
-        if (inputScale == 0) {
-            return num;
-        }
-        if (num < 0) {
-            return -roundShortDecimal(-num, rescaleFactor, inputScale);
+            long rescaleFactor = Decimals.longTenToNth(numScale.intValue());
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:715]: I think in short decimal it makes time difference. Just make sure that final time is better than original (no slowdown)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 8]
```diff
@@ -686,156 +698,109 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(B
...
     }
 
-    private static BigInteger roundLongDecimal(BigInteger num, BigInteger rescaleFactor)
-    {
-        BigInteger[] divideAndRemainder = num.divideAndRemainder(rescaleFactor);
-        BigInteger roundUp = divideAndRemainder[1].compareTo(rescaleFactor.shiftRight(1)) >= 0 ? ONE : ZERO;
-        return divideAndRemainder[0].add(roundUp);
-    }
+    @ScalarFunction("round")
+    @Description("round to nearest integer")
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: update comment

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 9]
```diff
@@ -686,156 +698,109 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(B
...
+        @LiteralParameters({"p", "s", "rp"})
+        @SqlType("decimal(rp, s)")
+        @Constraint(variable = "rp", expression = "min(38, p + 1)")
+        public static long roundNShort(
+                @LiteralParameter("p") Long numPrecision,
+                @LiteralParameter("s") Long numScale,
+                @SqlType("decimal(p, s)") long num,
+                @SqlType(BIGINT) long decimals)
+        {
+            if (num == 0 || numPrecision - numScale + decimals <= 0) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:748]: `numPrecision - numScale + decimals`: what case is this?
It only applies when `numPrecision - numScale = decimals = 0`. Please provide some explanation

@ilfrin

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 2 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:703]: Now `@LiteralParameter` work only with `Long` - not with `long`

`num` is OK. Scalar definitions for other types use `num`.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 3 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 7 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:715]: I will try to remember :)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 8 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 4 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:743]: no - this is consistent with other types

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 5 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:746]: ditto

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 9 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:748]: this is fine.
decimals can be negative

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 6 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:768]: ditto

File: presto-main/src/test/java/com/facebook/presto/type/TestDecimalOperators.java [hunk 10]
```diff
@@ -184,7 +184,7 @@ public void testDivide()
             throws Exception
     {
         // short short -> short
-        assertDecimalFunction("DECIMAL '1' / DECIMAL '2'", decimal("0"));
+        assertDecimalFunction("DECIMAL '1' / DECIMAL '3'", decimal("0"));
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/type/TestDecimalOperators.java]: shouldn't this be in @arhimondr commits?

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 11]
```diff
@@ -492,11 +494,7 @@ public WindowFunctionSupplier load(SpecializedFunctionKey key)
                 .function(TRY_CAST)
                 .function(DECIMAL_AVERAGE_AGGREGATION)
                 .function(DECIMAL_SUM_AGGREGATION)
-                .functions(DECIMAL_CEILING_FUNCTIONS)
-                .function(DECIMAL_FLOOR_FUNCTION)
-                .function(DECIMAL_MOD_FUNCTION)
-                .functions(DECIMAL_ROUND_FUNCTIONS)
-                .function(DECIMAL_TRUNCATE_FUNCTION);
+                .function(DECIMAL_MOD_FUNCTION);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:497]: don't add function at the end because this causes conflicts. Please leave TRY_CAST or DECIMAL_SUM_AGGREGATION or whatever there was last.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 12]
```diff
@@ -57,16 +61,31 @@
...
+        for (int scale = 1; scale < Decimals.MAX_PRECISION; ++scale) {
+            DECIMAL_HALF_UNSCALED_FOR_SCALE[scale] = UnscaledDecimal128Arithmetic.unscaledDecimal(
+                    BigInteger.TEN
+                            .pow(scale)
+                            .divide(BigInteger.valueOf(2)));
+            DECIMAL_ALMOST_HALF_UNSCALED_FOR_SCALE[scale] = UnscaledDecimal128Arithmetic.unscaledDecimal(
+                    BigInteger.TEN
+                            .pow(scale)
+                            .divide(BigInteger.valueOf(2))
+                            .subtract(BigInteger.valueOf(2)));
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: subtract ONE instead?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 13]
```diff
@@ -283,6 +255,44 @@ public static long ceilingFloat(@SqlType(StandardTypes.REAL) long num)
...
+        {
+            long rescaleFactor = Decimals.longTenToNth(numScale.intValue());
+            long increment = (num % rescaleFactor > 0) ? 1 : 0;
+            return num / rescaleFactor + increment;
+        }
+
+        @LiteralParameters({"p", "s", "rp"})
+        @SqlType("decimal(rp,0)")
+        @Constraint(variable = "rp", expression = "p - s + min(s, 1)")
+        public static Slice truncateLong(@LiteralParameter("s") Long numScale, @SqlType("decimal(p, s)") Slice num)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: why is it called truncate?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 14]
```diff
@@ -283,6 +255,44 @@ public static long ceilingFloat(@SqlType(StandardTypes.REAL) long num)
...
+            else {
+                tmp = add(num, DECIMAL_ALMOST_HALF_UNSCALED_FOR_SCALE[numScale.intValue()]);
+            }
+            return rescale(tmp, -numScale.intValue());
+        }
+
+        @LiteralParameters({"p", "s", "rp"})
+        @SqlType("decimal(rp,0)")
+        @Constraint(variable = "rp", expression = "p - s + min(s, 1)")
+        public static long truncateLongShort(@LiteralParameter("s") Long numScale, @SqlType("decimal(p, s)") Slice num)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 15]
```diff
@@ -283,6 +255,44 @@ public static long ceilingFloat(@SqlType(StandardTypes.REAL) long num)
...
+        public static Slice truncateLong(@LiteralParameter("s") Long numScale, @SqlType("decimal(p, s)") Slice num)
+        {
+            Slice tmp;
+            if (isNegative(num)) {
+                tmp = add(num, DECIMAL_HALF_UNSCALED_FOR_SCALE[numScale.intValue()]);
+            }
+            else {
+                tmp = add(num, DECIMAL_ALMOST_HALF_UNSCALED_FOR_SCALE[numScale.intValue()]);
+            }
+            return rescale(tmp, -numScale.intValue());
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:284]: why does it return Slice?

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 16]
```diff
@@ -170,14 +170,25 @@ public void testCeil()
...
         assertFunction("ceiling(DECIMAL '123')", createDecimalType(3), SqlDecimal.of("123"));
         assertFunction("ceiling(DECIMAL '-123')", createDecimalType(3), SqlDecimal.of("-123"));
+        assertFunction("ceiling(DECIMAL '123.00')", createDecimalType(4), SqlDecimal.of("123"));
+        assertFunction("ceiling(DECIMAL '-123.00')", createDecimalType(4), SqlDecimal.of("-123"));
         assertFunction("ceiling(DECIMAL '123.45')", createDecimalType(4), SqlDecimal.of("124"));
         assertFunction("ceiling(DECIMAL '-123.45')", createDecimalType(4), SqlDecimal.of("-123"));
+        assertFunction("ceiling(DECIMAL '123.49')", createDecimalType(4), SqlDecimal.of("124"));
+        assertFunction("ceiling(DECIMAL '-123.49')", createDecimalType(4), SqlDecimal.of("-123"));
+        assertFunction("ceiling(DECIMAL '123.50')", createDecimalType(4), SqlDecimal.of("124"));
+        assertFunction("ceiling(DECIMAL '-123.50')", createDecimalType(4), SqlDecimal.of("-123"));
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:193]: what about `123.01`, `123.99`, `-123.01`, `-123.99`?

Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:193]: add tests cases for near zero cases: `-0.49`, `-0.50`, `-0.01`

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 17]
```diff
@@ -324,13 +351,26 @@ public void testFloor()
 
         assertFunction("floor(DECIMAL '123')", createDecimalType(3), SqlDecimal.of("123"));
         assertFunction("floor(DECIMAL '-123')", createDecimalType(3), SqlDecimal.of("-123"));
+        assertFunction("floor(DECIMAL '123.00')", createDecimalType(4), SqlDecimal.of("123"));
+        assertFunction("floor(DECIMAL '-123.00')", createDecimalType(4), SqlDecimal.of("-123"));
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: ditto

File: presto-spi/src/main/java/com/facebook/presto/spi/function/Constraints.java [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+package com.facebook.presto.spi.function;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD})
+public @interface Constraints
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/Constraints.java:24]: unrelated changes

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 19]
```diff
@@ -1358,7 +1435,7 @@ private static void throwDivisionByZeroException()
         throw new PrestoException(DIVISION_BY_ZERO, "Division by zero");
     }
 
-    private static void incrementUnsafe(Slice decimal)
+    public static void incrementUnsafe(Slice decimal)
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: move above private methods?

File: presto-main/src/test/java/com/facebook/presto/type/TestDecimalOperators.java (hunk 10 above)
Comment (losipiuk) [reply presto-main/src/test/java/com/facebook/presto/type/TestDecimalOperators.java]: this was a mistake. Reverted

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java (hunk 11 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:497]: I just removed a bunch of functions.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 12 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 13 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 14 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 15 above)
Comment (losipiuk) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:284]: it is ok

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java (hunk 16 above)
Comment (losipiuk) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:193]: added

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java (hunk 17 above)
Comment (losipiuk) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java]: added

File: presto-spi/src/main/java/com/facebook/presto/spi/function/Constraints.java (hunk 18 above)
Comment (losipiuk) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/Constraints.java:24]: this is extracted as separate commit

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 19 above)
Comment (losipiuk) [reply presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: Extracted as separate commit and moved.

**Comment by losipiuk** [comment]
I addressed comments.
@pnowojski @sopel39 Is that ready to be merged to `td/feature_fast_decimal_v3`?
